### PR TITLE
MSPB-215: add support for teammate device

### PR DESCRIPTION
**Pull request by mavriciomart**
5.0 PR: https://github.com/2600hz/monster-ui-voip/pull/406
Ticket: https://2600hz-commercial.atlassian.net/browse/MSPB-215
Screenshots:
CleanShot 2022-03-17 at 17 40 06@2x
CleanShot 2022-03-17 at 17 40 49@2x
CleanShot 2022-03-17 at 17 41 20@2x
CleanShot 2022-03-17 at 17 42 12@2x
CleanShot 2022-03-17 at 17 42 29@2x

### CODE REVIEW

File: submodules/devices/devices.scss [hunk 1]
```diff
@@ -271,8 +272,15 @@
 	line-height: 75px;
 }
 
-.voip-edit-device-popup .edit-device .title-bar .device-title .device-icon > i {
-	font-size: 48px;
+.voip-edit-device-popup .edit-device .title-bar .device-title .device-icon > svg {
```
Comment (azefiel) [inline submodules/devices/devices.scss:275]: Looks like there is a sizing issue between these changes:
dev_add_device
and sandbox:
sandbox_add_device

File: submodules/devices/devices.scss [hunk 2]
```diff
@@ -271,8 +272,15 @@
...
+	width: 48px;
+	height: 48px;
+	vertical-align: middle;
+}
+
+.voip-edit-device-popup .edit-device .title-bar .device-title .device-icon > svg {
```
Comment (azefiel) [inline submodules/devices/devices.scss:281]: Isn't this a duplicate of the rule above? Maybe the last selector was meant to target an `i` tag?

File: submodules/devices/views/devices-teammate.html [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+					</div>
+					{{#if extra.hasDisabledRestrictions}}
+						{{#monsterText 'info' 'disabled-restrictions-info'}}
+							<p>{{ i18n.devices.popupSettings.restrictions.disabledRestrictionsInfo }}</p>
+						{{/monsterText}}
+					{{/if}}
+				</div>
+			</div>
+
+			{{#if showEmergencyCallerId}}
```
Comment (azefiel) [inline submodules/devices/views/devices-teammate.html:123]: From my understanding of the [spec](https://2600hz-commercial.atlassian.net/wiki/spaces/UCAAS/pages/1747616170/MS+Teams+Billing+Project#E911), this device should not expose e911 configuration at all, unless it changed since writing?

**Review by azefiel — COMMENTED** [review]
Looks great overall, minus some concerns regarding defaults definition and other minor typos/possible incorrect config.

It would have been awesome if this changes were broken down into several atomic commits vs one big one, to make the life of reviewers easier 🙏

File: submodules/devices/views/devices-teammate.html (hunk 3 above)
Comment (mavriciomart) [reply submodules/devices/views/devices-teammate.html:123]: I left the e911 configuration to follow the mockup from [this section](https://2600hz-commercial.atlassian.net/wiki/spaces/UCAAS/pages/1747616170/MS+Teams+Billing+Project#Adjustment-of-device-options-in-SmartPBX) in the documentation. Will ask for further clarification.

**Comment by mavriciomart** [comment]
@azefiel Thanks for the observations, I already addressed them and tried to separate my work in two commits.

File: submodules/devices/views/devices-teammate.html (hunk 3 above)
Comment (mavriciomart) [reply submodules/devices/views/devices-teammate.html:123]: From a conversation with Bret, we still do want to expose it.

**Review by azefiel — DISMISSED** [review]

File: submodules/devices/devices.js [hunk 4]
```diff
@@ -7,6 +7,10 @@ define(function(require) {
 		return _.every([dest, src], _.isArray) ? src : undefined;
 	};
 
+	var showTeammateDevice = monster.config.hasOwnProperty('extraDevices')
+		&& monster.config.extraDevices.length
+		&& $.inArray('teammate', monster.config.extraDevices) > -1;
+
```
Comment (azefiel) [inline submodules/devices/devices.js:13]: More declarative vs imperative:
```suggestion
	var showTeammateDevice = _
	    .chain(monster.config)
	    .get('extraDevices', [])
	    .includes('teammate')
	    .value();

```

File: submodules/users/users.js [hunk 5]
```diff
@@ -3763,7 +3767,8 @@ define(function(require) {
 				emptyAssigned: _.isEmpty(assigned),
 				emptySpare: _.isEmpty(unassigned),
 				assignedDevices: _.keyBy(assigned, 'id'),
-				unassignedDevices: _.keyBy(unassigned, 'id')
+				unassignedDevices: _.keyBy(unassigned, 'id'),
+				showTeammateDevice: showTeammateDevice
```
Comment (azefiel) [inline submodules/users/users.js:3771]: If you think about it, that new prop is not related to this function, in fact only the template cares about it.

How about inserting it as a template prop directly?

```js
template = $(self.getTemplate({
	name: 'devices',
	data: _,merge({
            showTeammateDevices: showTeammateDevices
        }, formattedResults),
	submodule: 'users'
}));
```

File: submodules/devices/devices.js [hunk 6]
```diff
@@ -58,7 +62,7 @@ define(function(require) {
 					'fax',
 					'ata',
 					'sip_uri',
-					'teammate'
+					showTeammateDevice ? 'teammate' : undefined
```
Comment (azefiel) [inline submodules/devices/devices.js:65]: This will potentially create an empty element, which in the grand scheme of things might not be a big deal BUT it would be nice to keep our data structures clean to avoid unexpected behavior down the road.

How about flattening nested arrays instead?

```js
addableDeviceTypes: _.flatten([[
      'sip_device',
      'cellphone',
      'smartphone',
      'softphone',
      'landline',
      'fax',
      'ata',
      'sip_uri'
  ],
  cond ? [
      'teammate'
  ] : []
])

**Review by azefiel — COMMENTED** [review]
Not sure what expectations are but, should devices of types listed in `extraDevices` be selectable/listable in the UI when that list is empty?

For example, `extraDevices=['teammate']` and I create a teammate device. Later, `extraDevices` gets updated and `teammate` removed. Should the UI still list those devices in the UI when fetching device documents?

**Comment by mavriciomart** [comment]
Thanks @azefiel! I addressed your comments in a newer commit.

> Not sure what expectations are but, should devices of types listed in extraDevices be selectable/listable in the UI when that list is empty?
For example, extraDevices=['teammate'] and I create a teammate device. Later, extraDevices gets updated and teammate removed. Should the UI still list those devices in the UI when fetching device documents?

The idea of this new key is that it will be set once the TeamMate integration has been configured for the cluster, and will allow the user to create a new teammate device.

I also updated the key name to `allowedExtraDeviceTypes` to be more clear of its purpose.

**Comment by mavriciomart** [comment]
@azefiel I got clarification from Karl on the above question.

> If they don't have teammate enabled but the system has devices of that type I would assume they must have enabled the device type at some point.  that means if they simply disappear they would have no way to remove them which means they have creds floating around they cant reset/remove (fraud / security risk) and no way to delete to avoid charges (billing).  So, yup the new parameter should only control showing the create option.

**Comment by azefiel** [comment]
Thanks for getting clarification on this @mavriciomart , that explanation is sound and makes total sense in hindsight.

File: submodules/devices/devices.js [hunk 7]
```diff
@@ -1218,7 +1230,19 @@ define(function(require) {
 						]))
 					},
 					smartphone: _.merge({}, sipSettings, callForwardSettings),
-					softphone: _.merge({}, sipSettings)
+					softphone: _.merge({}, sipSettings),
+					teammate: _.merge({
+						media: {
+							encryption: {
+								enforce_security: true,
+								webrtc: false,
```
Comment (harenson) [inline submodules/devices/devices.js:1238]: `webrtc` should go inside `media` not inside `media.encryption`.
